modules/revantageBidAdapter.js

```diff
@@ -0,0 +1,668 @@
+    if (canonical) {
+      metaTags.canonical = canonical.href;
+    }
+
+    return metaTags;
+  } catch (e) {
+    return {};
+  }
+}
+
+function getContentStructure() {
+  try {
+    if (typeof document === 'undefined') return {};
+
+    return {
+      // Headings for server-side analysis
+      headings: {
+        h1: getTextFromElements('h1'),
+        h2: getTextFromElements('h2', 5), // Limit to avoid bloat
+        h3: getTextFromElements('h3', 5)
+      },
+
+      // Sample content for server processing
+      contentSamples: {
+        firstParagraph: getFirstParagraphText(),
+        lastModified: document.lastModified || null
+      },
+
+      // Structured data (raw JSON for server parsing)
+      structuredData: getStructuredDataRaw(),
+
+      // Page elements that affect content type
+      pageElements: {
+        hasArticle: !!document.querySelector('article'),
+        hasVideo: !!document.querySelector('video'),
+        hasForm: !!document.querySelector('form'),
+        hasProduct: !!document.querySelector('[itemtype*="Product"]'),
+        hasBlog: !!document.querySelector('.blog, #blog, [class*="blog"]')
+      }
+    };
+  } catch (e) {
+    return {};
+  }
+}
+
+function getTextFromElements(selector, limit) {
+  limit = limit || 3;
+  try {
+    if (typeof document === 'undefined') return [];
+    const elements = document.querySelectorAll(selector);
+    const texts = [];
+
+    for (let i = 0; i < Math.min(elements.length, limit); i++) {
+      const text = elements[i].textContent;
+      if (text && text.trim().length > 0) {
+        texts.push(text.trim());
+      }
+    }
+    return texts;
+  } catch (e) {
+    return [];
+  }
+}
+
+function getFirstParagraphText() {
+  try {
+    if (typeof document === 'undefined') return '';
+    const firstP = document.querySelector('p');
+    if (firstP && firstP.textContent) {
+      return firstP.textContent.substring(0, 500); // Limit length
+    }
+    return '';
+  } catch (e) {
+    return '';
+  }
+}
+
+function getStructuredDataRaw() {
+  try {
+    if (typeof document === 'undefined') return [];
+    const scripts = document.querySelectorAll('script[type="application/ld+json"]');
+    const structuredData = [];
+
+    for (let i = 0; i < scripts.length; i++) {
+      try {
+        // Send raw JSON for server-side parsing
+        const data = JSON.parse(scripts[i].textContent);
+        structuredData.push(data);
+      } catch (e) {
+        // Invalid JSON, skip
+      }
+    }
+
+    return structuredData;
+  } catch (e) {
+    return [];
+  }
+}
+
+function getMediaElements() {
+  try {
+    if (typeof document === 'undefined') return {};
+
+    return {
+      imageCount: document.querySelectorAll('img').length,
+      videoCount: document.querySelectorAll('video').length,
+      iframeCount: document.querySelectorAll('iframe').length
+    };
+  } catch (e) {
+    return {};
+  }
+}
+
+// === ENHANCED DEVICE INFO ===
```

Comment:
Most of this should all be in the request object already, if not, let's put it there instead of in an adapter